lib/engine/game/g_18_fl/step/token.rb

```diff
@@ -42,19 +47,85 @@ def process_hex_token(action)
             token = action.token
 
             raise GameError, "#{hex.id} is not a town" if hex.tile.towns.empty?
-            raise GameError, "#{entity.name} already has a hotel in #{hex.tile.location_name}" if tokened(hex,
-                                                                                                          entity)
+            raise GameError, "#{entity.name} already has a hotel in #{hex.tile.location_name}" if tokened(hex, entity)
+
+            cost = action.cost # We are using token_cost_override
+            raise GameError, "#{entity.name} cannot afford "\
+                  "#{@game.format_currency(cost)} cost to lay hotel" if cost > entity.cash
 
-            @game.log << "#{entity.name} places a hotel on #{hex.name}"
+            @game.log << "#{entity.name} places a hotel on #{hex.name} for #{@game.format_currency(cost)}"
+            entity.spend(cost, @game.bank)
 
             entity.tokens.delete(token)
             hex.tile.icons << Part::Icon.new("../logos/18_fl/#{entity.id}")
             pass!
           end
 
+          def token_cost_override(entity, city_hex, _slot, token, _tokener)
+            hex = city_hex.respond_to?(:city?) ? city_hex.hex : city_hex
+            token.price * distance_to_station(entity, hex)
+          end
+
           def tokened(hex, entity)
             hex.tile.icons.any? { |i| i.name == entity.id }
           end
+
+          # How far is the start_hex from one of the corporation's stations?
+          # This is by track, not as the crow flies.
+          # start is either a city or hex (plain track or town)
+          def distance_to_station(corporation, start)
+            goal_hexes = corporation.tokens.select(&:city).map { |t| [t.city.hex, t.city] }
```

Comment:
I'm going to try re-implementing this w/ the second value I had `node.walk` yield; I think it might work but I'm not sure yet

Reply:
It works